### PR TITLE
Add new functions + tests

### PR DESCRIPTION
**Pull request by jacktweddle**
Added functions and tests for:

- minimum distance between a point and a plane
- central co-ords and radius of the circle when a plane intercepts a sphere
- equation of the line where two planes intersect
- point of intersection between a line segment and a plane
- rotation of a set of points about the x, y and z Cartesian axes
- translation of a point by a 3D vector
- converting between Cartesian, CPC and SPC coordinate systems

### CODE REVIEW

File: geom_utils.py [hunk 1]
```diff
@@ -59,12 +69,135 @@ def dist_between_planes(x1, y1, z1, d1, x2, y2, z2, d2):
...
+    plane = np.array([x, y, z, d])
+
+    # If val < 0, the two points must lie either side of the plane and so the line
+    # segment must intersect the plane
+    val = (np.dot(p0, plane[:3])-d)*(np.dot(p1, plane[:3])-d)
+
+    if val < 0:
+        if np.dot(p1-p0, plane[:3]) == 0:
+            return None  # Since line segment parallel to plane
+        t = t = plane[3] - np.dot(p0, plane[:3]) / np.dot(p1-p0, plane[:3])
```
Comment (py1sl) [inline geom_utils.py:130]: t = t =  
tidy up

File: geom_utils.py [hunk 2]
```diff
@@ -59,12 +69,135 @@ def dist_between_planes(x1, y1, z1, d1, x2, y2, z2, d2):
...
+
+    # To generate a point on the line, let z=z_ini - more flexible than just setting z=0
+
+    a = np.array([[x1, y1], [x2, y2]])
+    b = np.array([d1-(z1*z_ini), d2-(z2*z_ini)])
+    x = np.linalg.solve(a, b)
+
+    p = np.array([x[0], x[1], z_ini])
+
+    print(p, '+ t * ', n)
```
Comment (py1sl) [inline geom_utils.py:195]: remove or change to logging

File: geom_utils.py [hunk 3]
```diff
@@ -104,21 +237,75 @@ def coefficients_of_line_from_points(point_a, point_b):
...
-    qx = ox + np.cos(angle) * (px - ox) - np.sin(angle) * (py - oy)
-    qy = oy + np.sin(angle) * (px - ox) + np.cos(angle) * (py - oy)
-    return qx, qy
+    R = np.array([[1, 0, 0],
+                  [0, np.cos(theta), -np.sin(theta)],
+                  [0, np.sin(theta), np.cos(theta)]])
+
+    new_point_list = np.zeros(shape=(len(point_list), 3))
+
+    for i in range(len(point_list)):
```
Comment (py1sl) [inline geom_utils.py:253]: probably more pythonic to use 
```
for i, point in enumerate(point_list):
    new_point_list[i] = np.dot(R, (point.T - o.T))'
```

File: geom_utils.py [hunk 4]
```diff
@@ -104,21 +237,75 @@ def coefficients_of_line_from_points(point_a, point_b):
...
+    Returns array of rotated points
+    """
+
+    R = np.array([[np.cos(theta), 0, np.sin(theta)],
+                  [0, 1, 0],
+                  [-np.sin(theta), 0, np.cos(theta)]])
+
+    new_point_list = np.zeros(shape=(len(point_list), 3))
+
+    for i in range(len(point_list)):
```
Comment (py1sl) [inline geom_utils.py:272]: as above

File: geom_utils.py [hunk 5]
```diff
@@ -104,21 +237,75 @@ def coefficients_of_line_from_points(point_a, point_b):
...
+    Returns array of rotated points
+    """
+
+    R = np.array([[np.cos(theta), -np.sin(theta), 0],
+                  [np.sin(theta), np.cos(theta), 0],
+                  [0, 0, 1]])
+
+    new_point_list = np.zeros(shape=(len(point_list), 3))
+
+    for i in range(len(point_list)):
```
Comment (py1sl) [inline geom_utils.py:291]: as above

File: geom_utils.py [hunk 6]
```diff
@@ -104,21 +237,75 @@ def coefficients_of_line_from_points(point_a, point_b):
...
+
+def translate(point_list, translation):
+    """
+    Translates a point p by a translation vector
+    Points and translation vector should be given in the form of numpy arrays
+    """
+
+    new_point_list = np.zeros(shape=(len(point_list), 3))
+
+    for i in range(len(point_list)):
```
Comment (py1sl) [inline geom_utils.py:305]: as above

File: geom_utils.py [hunk 7]
```diff
@@ -292,3 +473,82 @@ def sphere_ray_intesect(x, y, z, ux, uy, uz, spherex, spherey,
...
+
+def cartesian_to_cylindrical(x, y, z):
+    """Converts Cartesian coordinates to cylindrical polar coordinates"""
+
+    rho = np.sqrt(x**2 + y**2)
+    if x >= 0:
+        phi = np.arctan(y/x)
+    elif x < 0:
+        phi = np.arctan(y/x) + np.pi
+    z_cyl = z
```
Comment (py1sl) [inline geom_utils.py:486]: remove, just return z

File: geom_utils.py [hunk 8]
```diff
@@ -292,3 +473,82 @@ def sphere_ray_intesect(x, y, z, ux, uy, uz, spherex, spherey,
...
     return (mu1, mu2)
+
+
+def cartesian_to_cylindrical(x, y, z):
+    """Converts Cartesian coordinates to cylindrical polar coordinates"""
+
+    rho = np.sqrt(x**2 + y**2)
+    if x >= 0:
+        phi = np.arctan(y/x)
+    elif x < 0:
```
Comment (py1sl) [inline geom_utils.py:484]: can just be else

File: geom_utils.py [hunk 9]
```diff
@@ -292,3 +473,82 @@ def sphere_ray_intesect(x, y, z, ux, uy, uz, spherex, spherey,
...
+
+
+def cartesian_to_spherical(x, y, z):
+    """Converts Cartesian coordinates to spherical polar coordinates"""
+
+    r = np.sqrt(x**2 + y**2 + z**2)
+    theta = np.arccos(z/r)
+    if x >= 0:
+        phi = np.arctan(y/x)
+    elif x < 0:
```
Comment (py1sl) [inline geom_utils.py:498]: can just be else

File: geom_utils.py [hunk 10]
```diff
@@ -292,3 +473,82 @@ def sphere_ray_intesect(x, y, z, ux, uy, uz, spherex, spherey,
...
+    return (r, theta, phi)
+
+
+def cylindrical_to_cartesian(rho, phi, z_cyl):
+    """Converts cylindrical polar coordinates to Cartesian coordinates"""
+
+    if rho < 0:
+        raise ValueError('Invalid input')
+
+    x = rho*np.cos(phi)
```
Comment (py1sl) [inline geom_utils.py:510]: pep8 spaces needed around operators

File: geom_utils.py [hunk 11]
```diff
@@ -292,3 +473,82 @@ def sphere_ray_intesect(x, y, z, ux, uy, uz, spherex, spherey,
...
+
+def spherical_to_cartesian(r, theta, phi):
+    """Converts spherical polar coordinates to Cartesian coordinates"""
+
+    if r < 0:
+        raise ValueError('Invalid input')
+
+    x = r*np.sin(theta)*np.cos(phi)
+    y = r*np.sin(theta)*np.sin(phi)
+    z = r*np.cos(theta)
```
Comment (py1sl) [inline geom_utils.py:525]: pep8 spaces needed around operators

File: geom_utils.py [hunk 12]
```diff
@@ -292,3 +473,82 @@ def sphere_ray_intesect(x, y, z, ux, uy, uz, spherex, spherey,
...
+
+def spherical_to_cylindrical(r, theta, phi):
+    """Converts spherical polar coordinates to cylindrical polar coordinates"""
+
+    if r < 0:
+        raise ValueError('Invalid input')
+
+    rho = r*np.sin(theta)
+    theta_cyl = phi
+    z_cyl = r*np.cos(theta)
```
Comment (py1sl) [inline geom_utils.py:538]: pep8 spaces needed around operators

File: tests/test_geom_util.py [hunk 13]
```diff
@@ -49,17 +49,81 @@ def test_vol_is_zero(self):
...
+        self.assertAlmostEqual(theta, np.pi/4)
+        self.assertAlmostEqual(z, 1)
+
+    def test_cartesian_to_spherical(self):
+        r, theta, phi = geom_utils.cartesian_to_spherical(1, 1, 1)
+        self.assertAlmostEqual(r, np.sqrt(3))
+        self.assertAlmostEqual(theta, np.arccos(1/np.sqrt(3)))
+        self.assertAlmostEqual(phi, np.pi/4)
+
+    def test_cylindrical_to_cartesian(self):
```
Comment (py1sl) [inline tests/test_geom_util.py:67]: should also test the error condition

File: tests/test_geom_util.py [hunk 14]
```diff
@@ -49,17 +49,81 @@ def test_vol_is_zero(self):
...
+        x, y, z = geom_utils.cylindrical_to_cartesian(np.sqrt(2), np.pi/4, 1)
+        self.assertAlmostEqual(x, 1)
+        self.assertAlmostEqual(y, 1)
+        self.assertAlmostEqual(z, 1)
+
+    def test_cylindrical_to_spherical(self):
```
Comment (py1sl) [inline tests/test_geom_util.py:73]: should also test the error condition

File: tests/test_geom_util.py [hunk 15]
```diff
@@ -49,17 +49,81 @@ def test_vol_is_zero(self):
...
+        r, theta, phi = geom_utils.cylindrical_to_spherical(np.sqrt(2), np.pi/4, 1)
+        self.assertAlmostEqual(r, np.sqrt(3))
+        self.assertAlmostEqual(theta, np.arccos(1/np.sqrt(3)))
+        self.assertAlmostEqual(phi, np.pi/4)
+
+    def test_spherical_to_cartesian(self):
```
Comment (py1sl) [inline tests/test_geom_util.py:79]: should also test the error condition

File: tests/test_geom_util.py [hunk 16]
```diff
@@ -49,17 +49,81 @@ def test_vol_is_zero(self):
...
+        x, y, z = geom_utils.spherical_to_cartesian(np.sqrt(3), np.arccos(1/np.sqrt(3)), np.pi/4)
+        self.assertAlmostEqual(x, 1)
+        self.assertAlmostEqual(y, 1)
+        self.assertAlmostEqual(z, 1)
+
+    def test_spherical_to_cylindrical(self):
```
Comment (py1sl) [inline tests/test_geom_util.py:85]: should also test the error condition

**Review by py1sl — CHANGES_REQUESTED** [review]
generally looks good, few minor changes to make.  testing could be expanded, 
nice use of np.testing for arrays, I've not tried that before

**Comment by jacktweddle** [comment]
All the requested changes have been made, I think it automatically updates here?

**Review by py1sl — APPROVED** [review]

**Comment by jacktweddle** [comment]
Hi Steve,

I've updated all the functions to use numpy arrays and have added new tests that have improved the coverage to 98%. Couldn't quite get 100% because of a few logging cases :(

Hopefully all is good!

**Comment by jacktweddle** [comment]
pep8 issues have now been cleared up :)

**Pull request merged**